conda/match.py

```diff
@@ -0,0 +1,203 @@
+                        field, eq, value = (x.strip() for x in opart.partition('='))
+                        if not field:
+                            continue
+                        elif not value and (eq or field != 'optional'):
+                            raise CondaValueError("Invalid MatchSpec: %s" % spec)
+                        elif field == 'optional':
+                            kwargs.setdefault('optional', bool(value) if eq else True)
+                        elif field == 'target':
+                            kwargs.setdefault('target', value)
+                        else:
+                            push_((field, literal_eval(value)))
+
+            else:
+                raise CondaValueError("Cannot construct MatchSpec from: %r" % (spec,))
+
+        self.target = kwargs.pop('target', None)
+        self.optional = bool(kwargs.pop('optional', False))
+        push_(*iteritems(kwargs))
+        return self
+
+    def exact_field(self, fn):
+        v = self.specs.get(fn)
+        return None if v is None or hasattr(v, 'match') else v
+
+    def is_exact(self):
+        return all(self.exact_field(x) is not None for x in ('fn', 'schannel'))
+
+    def is_simple(self):
+        return len(self.specs) == 1 and self.exact_field('name') is not None
+
+    def match(self, rec):
+        # type: (IndexRecord) -> bool
+        for f, v in iteritems(self.specs):
+            val = getattr(rec, f)
+            if not (v.match(val) if hasattr(v, 'match') else v == val):
+                return False
+        return True
+
+    def to_filename(self):
+        tmp = self.exact_field('fn')
+        if tmp:
+            return tmp
+        vals = [self.exact_field(x) for x in ('name', 'version', 'build')]
+        if not any(x is None for x in vals):
+            return '%s-%s-%s.tar.bz2' % tuple(map(str, vals))
+
+    def to_dict(self, args=True):
+        res = self.specs.copy()
+        if args and self.optional:
+            res['optional'] = bool(self.optional)
+        if args and self.target is not None:
+            res['target'] = self.target
+        return res
+
+    def to_string(self, args=True, base=True):
+        nf = (3 if 'build' in self.specs else
+              (2 if 'version' in self.specs else 1)) if base else 0
+        flds = ('name', 'version', 'build')[:nf]
+        base = ' '.join(str(self.specs.get(f, '*')) for f in flds)
+        xtra = ['%s=%r' % (f, v) for f, v in sorted(iteritems(self.specs))
+                if f not in flds]
+        if args and self.optional:
+            xtra.append('optional' if base else 'optional=True')
+        if args and self.target:
+            xtra.append('target=' + self.target)
+        xtra = ','.join(xtra)
+        if not base:
+            return xtra
+        elif xtra:
+            return '%s (%s)' % (base, xtra)
+        else:
+            return base
+
+    def __eq__(self, other):
+        if isinstance(other, MatchSpec):
```

Comment:
I think `__eq__` should always return True or False.  At least I'm more comfortable that way.

so just

    return isinstance(other, Matchspec) and ((self.specs, self.optional, self.target) == (other.specs, other.optional, other.target))